apps/communications/contacts/views/form/js/form_ui.js

```diff
@@ -0,0 +1,790 @@
+
+          var type = target.dataset.type;
+          var selector = target.dataset.selector;
+
+          var container = configs[type].container;
+          var elem = document.getElementById(selector);
+
+          if (type !== 'photo') {
+            // Remove field
+            elem.parentNode.removeChild(elem);
+            // Show empty if no more fields of the same type are available
+            if (container.querySelectorAll('[data-field]').length === 0) {
+              container.classList.add('empty');
+            }
+          }
+
+          counters[type]--;
+          // In this version only two dates are allowed
+          if (type === 'date') {
+            checkAddDateButton();
+          }
+
+          cachedInputs = null;
+          // textFieldsCache.clear();
+          checkDisableButton();
+          break;
+      }
+    });
+
+    thumbAction.addEventListener(touchstart, function click(event) {
+      // Removing current photo
+      if (event.target.tagName === 'BUTTON') {
+        saveButton.removeAttribute('disabled');
+      }
+    });
+
+    // 'ValueModified' comes from the 'TagSelector', and will return a value
+    // from a set of Tags.
+    formView.addEventListener('ValueModified', function onValueModified(event) {
+      if (!event.detail) {
+        return;
+      }
+
+      if (!emptyForm() && event.detail.prevValue !== event.detail.newValue) {
+        saveButton.removeAttribute('disabled');
+      }
+    });
+
+    formHeader.addEventListener('action', function() {
+      window.dispatchEvent(new Event('close-ui'));
+    });
+
+    saveButton.addEventListener('click', function() {
+      var contact = {};
+      var inputs = { givenName, familyName };
+
+      for (var field in inputs) {
+        var value = inputs[field].value;
+        if (value && value.length > 0) {
+          contact[field] = [value];
+        } else {
+          contact[field] = null;
+        }
+      }
+
+      if (company.value && company.value.length > 0) {
+        contact.org = [company.value];
+      }
+
+      contact.email = getEmails();
+      contact.tel = getPhones();
+      contact.adr = getAddresses();
+      contact.note = getNotes();
+      var dates = getDates();
+      contact.bday = dates.bday;
+      contact.anniversary = dates.anniversary;
+      contact.name = getName(contact);
+
+      if (currentPhoto) {
+        contact.photo = [currentPhoto];
+      }
+
+      var customEvent = new CustomEvent('save-contact', {'detail': contact});
+      window.dispatchEvent(customEvent);
+    });
+
+    contactForm.addEventListener('click', function(e) {
+      e.preventDefault();
+      e.stopPropagation();
+
+      if (e.target.dataset && e.target.dataset.action === 'delete') {
+        return;
+      }
+      if (e.target.dataset && e.target.dataset.fieldType) {
+        var type = e.target.dataset.fieldType;
+        insertField(
+          type,
+          null,
+          true
+        );
+      }
+    });
+  }
```

Comment:
In general I like to have the functions defined and then attaching the events, so we can have documentation for the functions like:

```
formHeader.addEventListener('action', closeForm);
saveButton.addEventListener('click', saveContact);
contactForm.addEventListener('click', addField);
```